internal/hierarchyconfig/reconciler.go

```diff
@@ -189,7 +189,7 @@ func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, nm string)
 	r.updateFinalizers(log, inst, nsInst, anms)
 
 	// Sync the Hierarchy singleton with the in-memory forest.
-	needUpdateObjects := r.syncWithForest(log, nsInst, inst, deletingCRD, anms)
+	needUpdateObjects := r.syncWithForest(ctx, log, nsInst, inst, deletingCRD, anms)
```

Comment:
I've generally been trying to ensure that there are no calls to the client in `syncWithForest` because the mutex is being held and I want to be certain we don't make a call to the apiserver. I've recently learned that the caching is probably more guaranteed than I thought but I'm not certain, so I think I'd like to preserve this behaviour for now.

Instead, if you look up a few lines, you'll see the comment `Get a list of subnamespace anchors from apiserver`. We could move the code from`syncSubnamespaceParent` to here, and just pass the anchor (if it exists) down into `syncWithForest`. It's a bit messy but I'd prefer that for architectural reasons for now.

```
anchor, err := r.getSubnamespaceAnchorInstance(ctx, nsInst.Annotations[api.SubnamespaceOf], nm)
if err != nil && !errors.IsNotFound(err) {
  return err
}
```

This actually probably allows you to get rid of the loop in `syncSubnamespaceParent` as well, since we'll already have checked the apiserver for the anchor and will know if it's missing.

Reply:
Hi. I've pushed up changes based on this. All the tests still pass and the logic still makes sense to me. Would be good if you could confirm the changes are what you expected.